File: android/app/src/main/java/fr/magiclockscreen/android/MagicWallpaperClient.java
package fr.magiclockscreen.android;

import android.app.WallpaperManager;
import android.content.Context;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;

import java.io.ByteArrayOutputStream;
import java.io.InputStream;
import java.net.HttpURLConnection;
import java.net.URL;
import java.net.URLEncoder;
import java.nio.charset.StandardCharsets;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

public final class MagicWallpaperClient {
    private MagicWallpaperClient() {}

    public static final class MagicResult {
        public final String value;
        public final String hash;

        public MagicResult(String value, String hash) {
            this.value = value;
            this.hash = hash;
        }
    }

    private static String enc(String value) throws Exception {
        return URLEncoder.encode(value, "UTF-8");
    }

    private static String getText(String url) throws Exception {
        HttpURLConnection c = (HttpURLConnection) new URL(url).openConnection();
        c.setConnectTimeout(10000);
        c.setReadTimeout(20000);
        c.setRequestMethod("GET");
        c.setRequestProperty("User-Agent", "MagicLockscreenAndroid/0.3");
        int code = c.getResponseCode();
        InputStream stream = code >= 200 && code <= 299 ? c.getInputStream() : c.getErrorStream();
        String text = new String(readAll(stream), StandardCharsets.UTF_8);
        if (code < 200 || code > 299) throw new IllegalStateException("HTTP " + code + " : " + text);
        return text;
    }

    private static byte[] getBytes(String url) throws Exception {
        HttpURLConnection c = (HttpURLConnection) new URL(url).openConnection();
        c.setConnectTimeout(10000);
        c.setReadTimeout(30000);
        c.setRequestMethod("GET");
        c.setRequestProperty("User-Agent", "MagicLockscreenAndroid/0.3");
        int code = c.getResponseCode();
        if (code < 200 || code > 299) {
            String err = c.getErrorStream() == null ? "" : new String(readAll(c.getErrorStream()), StandardCharsets.UTF_8);
            throw new IllegalStateException("Image HTTP " + code + " : " + err);
        }
        return readAll(c.getInputStream());
    }

    private static byte[] readAll(InputStream stream) throws Exception {
        ByteArrayOutputStream out = new ByteArrayOutputStream();
        byte[] buffer = new byte[8192];
        int n;
        while ((n = stream.read(buffer)) != -1) out.write(buffer, 0, n);
        stream.close();
        return out.toByteArray();
    }

    private static String jsonString(String json, String key) {
        Pattern p = Pattern.compile("\\\"" + Pattern.quote(key) + "\\\"\\s*:\\s*\\\"((?:\\\\.|[^\\\"])*)\\\"");
        Matcher m = p.matcher(json);
        if (!m.find()) return "";
        return m.group(1).replace("\\\"", "\"").replace("\\n", "\n").replace("\\/", "/");
    }

    public static MagicResult fetchValue(String backend, String token) throws Exception {
        String base = backend == null ? "" : backend.trim();
        while (base.endsWith("/")) base = base.substring(0, base.length() - 1);
        String json = getText(base + "/api/value?token=" + enc(token));
        if (!json.contains("\"ok\"") || !json.contains("true")) throw new IllegalStateException("Réponse backend invalide : " + json);
        String value = jsonString(json, "value");
        String hash = jsonString(json, "hash");
        if (hash.length() == 0) hash = value;
        if (value.trim().length() == 0) throw new IllegalStateException("Aucune révélation trouvée");
        return new MagicResult(value, hash);
    }

    public static MagicResult updateLockscreen(Context context, String backend, String token) throws Exception {
        MagicResult result = fetchValue(backend, token);
        String base = backend == null ? "" : backend.trim();
        while (base.endsWith("/")) base = base.substring(0, base.length() - 1);
        byte[] bytes = getBytes(base + "/api/wallpaper?token=" + enc(token) + "&t=" + System.currentTimeMillis());
        Bitmap bitmap = BitmapFactory.decodeByteArray(bytes, 0, bytes.length);
        if (bitmap == null) throw new IllegalStateException("Image illisible");
        WallpaperManager.getInstance(context).setBitmap(bitmap, null, true, WallpaperManager.FLAG_LOCK);
        return result;
    }
}

File: android/app/src/main/java/fr/magiclockscreen/android/MagicWallpaperClient.kt
package fr.magiclockscreen.android

import android.app.WallpaperManager
import android.content.Context
import android.graphics.BitmapFactory
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import org.json.JSONObject
import java.io.BufferedInputStream
import java.net.HttpURLConnection
import java.net.URL
import java.net.URLEncoder
import java.nio.charset.StandardCharsets

data class MagicValue(
    val value: String,
    val hash: String
)

object MagicWallpaperClient {
    suspend fun fetchValue(backendBaseUrl: String, token: String): MagicValue = withContext(Dispatchers.IO) {
        val url = buildUrl(backendBaseUrl, "/api/value", token)
        val text = httpGetText(url)
        val json = JSONObject(text)
        if (!json.optBoolean("ok", false)) {
            throw IllegalStateException(json.optString("error", "Erreur valeur"))
        }
        MagicValue(
            value = json.optString("value", "").trim(),
            hash = json.optString("hash", "").trim()
        )
    }

    suspend fun updateLockscreen(context: Context, backendBaseUrl: String, token: String): MagicValue = withContext(Dispatchers.IO) {
        val value = fetchValue(backendBaseUrl, token)
        if (value.value.isBlank()) throw IllegalStateException("Révélation vide")

        val wallpaperUrl = buildUrl(backendBaseUrl, "/api/wallpaper", token)
        val bitmap = httpGetBitmap(wallpaperUrl)
        val manager = WallpaperManager.getInstance(context.applicationContext)

        manager.setBitmap(
            bitmap,
            null,
            true,
            WallpaperManager.FLAG_LOCK
        )

        value
    }

    private fun buildUrl(base: String, path: String, token: String): String {
        val cleanBase = base.trim().trimEnd('/')
        val encodedToken = URLEncoder.encode(token.trim(), StandardCharsets.UTF_8.toString())
        return "$cleanBase$path?token=$encodedToken"
    }

    private fun httpGetText(url: String): String {
        val connection = (URL(url).openConnection() as HttpURLConnection).apply {
            requestMethod = "GET"
            connectTimeout = 12_000
            readTimeout = 20_000
            setRequestProperty("Accept", "application/json")
            useCaches = false
        }
        return try {
            val code = connection.responseCode
            val stream = if (code in 200..299) connection.inputStream else connection.errorStream
            val text = stream.bufferedReader().use { it.readText() }
            if (code !in 200..299) throw IllegalStateException("HTTP $code : $text")
            text
        } finally {
            connection.disconnect()
        }
    }

    private fun httpGetBitmap(url: String) = with(URL(url).openConnection() as HttpURLConnection) {
        requestMethod = "GET"
        connectTimeout = 12_000
        readTimeout = 30_000
        setRequestProperty("Accept", "image/png,image/*")
        useCaches = false
        try {
            val code = responseCode
            if (code !in 200..299) {
                val text = errorStream?.bufferedReader()?.use { it.readText() } ?: ""
                throw IllegalStateException("HTTP $code wallpaper : $text")
            }
            BufferedInputStream(inputStream).use { input ->
                BitmapFactory.decodeStream(input) ?: throw IllegalStateException("Image wallpaper illisible")
            }
        } finally {
            disconnect()
        }
    }
}
